pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1345 @@
+	/// (Past and present rounds are unaffected.)
+	///
+	/// **Warning:** Fallible function that changes storage.
+	fn update_future_rounds(
+		&mut self,
+		payment: BalanceOf<T>,
+		submission_count_bounds: (u32, u32),
+		restart_delay: T::RoundId,
+		timeout: T::BlockNumber,
+	) -> DispatchResult {
+		let (min, max) = submission_count_bounds;
+		ensure!(max >= min, Error::<T>::WrongBounds);
+		// Make sure that both the min and max of submissions is
+		// less or equal to the number of oracles.
+		ensure!(self.oracle_count() >= max, Error::<T>::MaxExceededTotal);
+		// Make sure that at least one oracle can request a new
+		// round.
+		ensure!(
+			self.oracle_count() as u64 > restart_delay.into(),
+			Error::<T>::DelayExceededTotal
+		);
+		// require(recordedFunds.available >= requiredReserve(_paymentAmount), "insufficient funds for payment");
+		if self.oracle_count() > 0 {
+			ensure!(min > 0, Error::<T>::WrongBounds);
+		}
+
+		self.config.payment = payment;
+		self.config.submission_count_bounds = submission_count_bounds;
+		self.config.restart_delay = restart_delay;
+		self.config.timeout = timeout;
+
+		Module::<T>::deposit_event(RawEvent::RoundDetailsUpdated(
+			self.id,
+			payment,
+			submission_count_bounds,
+			restart_delay,
+			timeout,
+		));
+		Ok(())
+	}
+
+	/// Initialize a new round.
+	/// Will close the previous one if it is timed out.
+	///
+	/// **Warning:** Fallible function that changes storage.
+	fn initialize_round(
+		&self,
+		new_round_id: T::RoundId,
+	) -> Result<T::BlockNumber, DispatchError> {
+		let prev_round_id = new_round_id.saturating_sub(One::one());
+		if self.is_timed_out(prev_round_id) {
+			self.close_timed_out_round(prev_round_id)?;
+		}
+
+		Details::<T>::insert(
+			self.id,
+			new_round_id,
+			RoundDetails {
+				submissions: Vec::new(),
+				submission_count_bounds: self.config.submission_count_bounds,
+				payment: self.config.payment,
+				timeout: self.config.timeout,
+			},
+		);
+		let started_at = frame_system::Module::<T>::block_number();
+		Rounds::<T>::insert(self.id, new_round_id, Round::new(started_at));
+
+		Ok(started_at)
+	}
+
+	/// Close a timed out round and remove its details.
+	///
+	/// **Warning:** Fallible function that changes storage.
+	fn close_timed_out_round(&self, timed_out_id: T::RoundId) -> DispatchResult {
+		let prev_id = timed_out_id.saturating_sub(One::one());
+		let prev_round = self.round(prev_id).ok_or(Error::<T>::RoundNotFound)?;
+		let mut timed_out_round = self.round(timed_out_id).ok_or(Error::<T>::RoundNotFound)?;
+		timed_out_round.answer = prev_round.answer;
+		timed_out_round.answered_in_round = prev_round.answered_in_round;
+		let updated_at = frame_system::Module::<T>::block_number();
+		timed_out_round.updated_at = Some(updated_at);
+
+		Rounds::<T>::insert(self.id, timed_out_id, timed_out_round);
+		Details::<T>::remove(self.id, timed_out_id);
+
+		Ok(())
+	}
+
+	/// Store the feed config in storage.
+	fn sync_to_storage(&mut self) {
```

Comment:
this is a pretty good pattern, I think we can introduce it in substrate as well (types that automatically go into storage upon drop).

Reply:
but be wary of bugs as well.. if you miss flipping `should_sync`..

Reply:
The pattern I use here is that I initialize the interface with one intention (either drop the interface OR sync to storage on drop) and then don't alter `should_sync`.

Reply:
I'd actually like it to be constant for a particular interface, but creating different types would be complicated :-/

Reply:
I ended up separating the interfaces to track the different sync to storage behaviors